### PR TITLE
Fix scatter and gather op when has duplicate index

### PR DESCRIPTION
**Pull request by wawltor**
Fix #https://github.com/PaddlePaddle/Paddle/issues/17509
The scatter op has a calc bug when the indices has same index, the scatter op use overwrite mode to calculate the same index, fix this bug by using the accumulate mode to calculate the same index.At the same time, the gather op has the same bug when the op calc the grad. And we use the lib of open-blas and eigen to optimize the time cost in accumulate mode.

### CODE REVIEW

File: paddle/fluid/operators/gather_op.cc [hunk 1]
```diff
@@ -74,6 +74,13 @@ class GatherOpMaker : public framework::OpProtoAndCheckerMaker {
     AddInput("X", "The source input of gather op");
     AddInput("Index", "The index input of gather op");
     AddOutput("Out", "The output of gather op");
+    AddAttr<bool>(
+        "overwrite",
+        "(bool, defalut: False) "
```
Comment (xsrobin) [inline paddle/fluid/operators/gather_op.cc]: defalut->default

File: paddle/fluid/operators/scatter.h [hunk 2]
```diff
@@ -14,17 +14,55 @@ limitations under the License. */
...
 #include "paddle/fluid/platform/place.h"
+#include "unordered_set"
 
 namespace paddle {
 namespace operators {
 
 using Tensor = framework::Tensor;
 
+/**
+  * Return the updated array pointer, use blas or eigen lib to optimize time
```
Comment (xsrobin) [inline paddle/fluid/operators/scatter.h]: blas->bias? eigen lib？

File: paddle/fluid/operators/scatter.h [hunk 3]
```diff
@@ -14,17 +14,55 @@ limitations under the License. */
...
+  auto src_slice = src.Slice(src_index, src_index + 1);
+  auto dist_slice = dist->Slice(dist_index, dist_index + 1);
+
+  auto eigen_src = framework::EigenVector<T>::Flatten(src_slice);
+  auto eigen_dist = framework::EigenVector<T>::Flatten(dist_slice);
+
+  eigen_dist += eigen_src;
+}
 /**
  * Return a updated tensor from source tensor, scattered according to index:
```
Comment (xsrobin) [inline paddle/fluid/operators/scatter.h]: an updated

File: paddle/fluid/operators/scatter.h [hunk 4]
```diff
@@ -64,5 +102,47 @@ void ScatterAssign(const platform::DeviceContext& ctx, const Tensor& src,
...
+  for (int i = 1; i < src_dims.size(); i++)
+    PADDLE_ENFORCE(src_dims[i] == dst_dims[i]);
+
+  // slice size
+  size_t slice_size = 1;
+  for (int i = 1; i < src_dims.size(); ++i) slice_size *= src_dims[i];
+
+  const size_t& slice_bytes = slice_size * sizeof(T);
+
+  // if in not overwrite mode, need to init output data
```
Comment (xsrobin) [inline paddle/fluid/operators/scatter.h]: if not in

**Review by xsrobin — COMMENTED** [review]
as comments

**Review by xsrobin — APPROVED** [review]
LGTM

**Review by kuke — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]

**Pull request merged**